components/bitbucket/bitbucket.app.js

```diff
@@ -0,0 +1,150 @@
+const axios = require("axios");
+
+module.exports = {
+  type: "app",
+  app: "bitbucket",
+  propDefinitions: {
+    workspaceId: {
+      type: "string",
+      label: "Workspace",
+      description: "The workspace that contains the repositories to work with.",
+      async options(context) {
+        const url = this._userWorkspacesEndpoint();
+        const params = {
+          sort: "name",
+          fields: [
+            "values.name",
+            "values.slug",
+          ],
+        };
+
+        const data = await this._propDefinitionsOptions(url, params, context);
+        const options = data.values.map(workspace => ({
+          label: workspace.name,
+          value: workspace.slug,
+        }));
+        return {
+          options,
+          context: {
+            // https://developer.atlassian.com/bitbucket/api/2/reference/meta/pagination
+            nextPageUrl: data.next,
+          },
+        };
+      },
+    },
+    repositoryId: {
+      type: "string",
+      label: "Repository",
+      description: "The repository for which the events will be processed.",
+      async options(context) {
+        const { workspaceId } = context;
+        const url = this._workspaceRepositoriesEndpoint(workspaceId);
+        const params = {
+          sort: "slug",
+          fields: [
+            "values.slug",
+          ],
+        };
+
+        const data = await this._propDefinitionsOptions(url, params, context);
+        const options = data.values.map(repository => ({
+          label: repository.slug,
+          value: repository.slug,
+        }));
+        return {
+          options,
+          context: {
+            // https://developer.atlassian.com/bitbucket/api/2/reference/meta/pagination
+            nextPageUrl: data.next,
+          },
+        };
+      },
+    },
+  },
+  methods: {
+    _apiUrl() {
+      return "https://api.bitbucket.org/2.0";
+    },
+    _userWorkspacesEndpoint() {
+      const baseUrl = this._apiUrl();
+      return `${baseUrl}/workspaces`;
+    },
+    _workspaceRepositoriesEndpoint(workspaceId) {
+      const baseUrl = this._apiUrl();
+      return `${baseUrl}/repositories/${workspaceId}`;
+    },
+    _hooksEndpointUrl(workspaceId, repositoryId) {
+      const baseUrl = this._workspaceRepositoriesEndpoint(workspaceId);
+      return `${baseUrl}/${repositoryId}/hooks`;
+    },
+    _hookEndpointUrl(workspaceId, repositoryId, hookId) {
+      const baseUrl = this._hooksEndpointUrl(workspaceId, repositoryId);
+      // https://developer.atlassian.com/bitbucket/api/2/reference/meta/uri-uuid#uuid
+      return `${baseUrl}/{${hookId}}`;
+    },
+    async _propDefinitionsOptions(url, params, { page, prevContext }) {
+      let requestConfig = this._makeRequestConfig();  // Basic axios request config
+      if (page === 0) {
+        // First time the options are being retrieved.
+        // Include the parameters provided, which will be persisted
```

Comment:
Do you just mean the nextPageUrl cursor will implicitly include the initial params?

Reply:
(nitpicky, but either way this comment is a little unclear)

Reply:
Yes, that's exactly what I meant :slightly_smiling_face: 
I'll create a separate PR to rephrase it (similar logic with the same comment is in other components as well)

Reply:
Added a task for this: https://github.com/PipedreamHQ/pipedream/issues/508

Reply:
Created task for this: https://github.com/PipedreamHQ/pipedream/issues/509